qiskit_experiments/library/characterization/analysis/resonator_spectroscopy_analysis.py

```diff
@@ -0,0 +1,71 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Spectroscopy analysis class for resonators."""
+
+from typing import List, Tuple
+import numpy as np
+
+import qiskit_experiments.curve_analysis as curve
+from qiskit_experiments.curve_analysis import ResonanceAnalysis
+from qiskit_experiments.framework import AnalysisResultData, ExperimentData
+from qiskit_experiments.framework.matplotlib import get_non_gui_ax
+from qiskit_experiments.data_processing.processor_library import ProjectorType
+
+
+class ResonatorSpectroscopyAnalysis(ResonanceAnalysis):
+    """Class to analysis resonator spectroscopy."""
+
+    @classmethod
+    def _default_options(cls):
+        options = super()._default_options()
+        options.dimensionality_reduction = ProjectorType.ABS
+        options.result_parameters = [curve.ParameterRepr("freq", "meas_freq", "Hz")]
+        options.plot_iq_data = True
+        return options
+
+    def _run_analysis(
+        self, experiment_data: ExperimentData
+    ) -> Tuple[List[AnalysisResultData], List["pyplot.Figure"]]:
+        """Wrap the analysis to optionally plot the IQ data."""
+        analysis_results, figures = super()._run_analysis(experiment_data)
+
+        if self.options.plot_iq_data:
+            axis = get_non_gui_ax()
+            figure = axis.get_figure()
+            figure.set_size_inches(*self.options.style.figsize)
+
+            iqs = []
+
+            for datum in experiment_data.data():
+                if "memory" in datum:
+                    mem = np.array(datum["memory"])
+
+                    # Average single-shot data.
+                    if len(mem.shape) == 3:
+                        iqs.append(np.average(mem.reshape(mem.shape[0], mem.shape[2]), axis=0))
```

Comment:
That works if shape is (X, 1, Y) as you correctly point out. This is perhaps a bit more robust: 7c44dd1

Reply:
Oh, I see. I couldn't remember which dimension was which. So the dimensions are `(shot #, slot #, quadrature)`, right? And we are only expecting a single measurement slot so we are dropping the slot dimension. What about just this?

    iqs.append(np.average(mem[:, 0, :], axis=0))

Your suggestion might also be fine if we ever wanted to put in multiple acquisitions in a single circuit, which for simple resonator spectroscopy might be better than taking more shots in some cases. Otherwise though it might be confusing since usually multiple slots implies multiple qubits.

I really wish the qiskit data containers used more typing or somehow made it easier to track what shape the data was in.